### ISSUE TITLE
Separate model from view

### ISSUE THREAD

**Issue by talesm** [TalesM/orcamento#3]
Having the model and the view coupled is bad and I should feel bad. Should decouple it in sort of document/view or MVC scheme.
# Occurrences
- OrcamentoMainFrame;
- ExecutionDialog;
- WalletOverview;

**Comment by talesm** [comment]
Updating on current status, I created a local branch to do the refactoring to support wx's document/view framework. I failed to accomplish that due my inexperience in this framework and the horrible support for it on wxSmith (i.e., no support). So I got tired to fight the framework and the IDE at same time and threw away that branch.

Then I decided replicate the framework features. I is not that hard as I did not want multiple documents open at same time nor other features closely related to the the ui. I then compiled my need in three features:
1. Decoupling the visualization (the selects);
2. Decoupling the manipulation (change commands);
3. The undo/redo support;
4. Ability to control the lifetime;

I then designed a simple way decouple, as it fall in simple design patterns: The 1 is a case of [iterator](http://en.wikipedia.org/wiki/Iterator_pattern) (I also used the [template](http://en.wikipedia.org/wiki/Template_method_pattern) to reuse common commands, but it still need some love). The commands, is even more obvious, as it is the exact name of the [pattern](http://en.wikipedia.org/wiki/Command_pattern).

So I created the OrcaDocument class, to abstract away the database. Then the template class OrcaView is base for different views (bad name, it is no related to sql views, but I can't think any better). Each OrcaView implementation acts as an iterator for the database. I didn't follow the C++ way (2 _pointers_), but it is iterated through the execution of a lambda. The result is very close to what a for-each would look, and is very simpler IMO. 

The number 2 is very classic Command, as said before, and it is easily implemented with an abstract OrcaAction and multiple derived ones. However the undo feature needs some thought to get correctly handling of ui changes, so even though we could implement undo at model level I decided to postpone the number 3 to #12.

The #4 could be done with a simple manipulation of transactions. But as it changes deeply the work-flow I choose to postpone to #11.

Currently I'm still doing the number two, as there is many actions to map.

**Comment by talesm** [comment]
Finally I merged back the decoupling branch. All actions were found and turned into subclass of OrcaAction. I tried to put a sensible hierarchy, but I'm not 100% with it. The problem is the estimate's and executions update actions: all other actions were a one-to-one map to its equivalent sql command, but these have instead a template base class (UpdateEstimateBase and UpdateExecutionBase) and as many subclass of it as there is mutable fields (e.g UpdateEstimateName, UpdateExecutionDescription and so on). 

I feel I could avoid this third generation through the use of some template tricks (some trait-like pattern) and have a simpler hierarchy. But it would took more time to figure out how to do it and _I really abused template cleverness_ on OrcaView hierarchy, so I stick with the time table and give a rest to the gcc (it already suffers for OrcaView). It's possible I redo it someday, but it is a lower priority as task now as I already got working code.
